### PR TITLE
convert zircon and fuchsia to null-safety

### PR DESCRIPTION
**Pull request by chaselatta**
## Description

Changes the `--nnbd-weak` flag to `--nnbd-agnostic` and migrates the native zircon and fuchsia packages to be null safe. This will allow us to start using nullability features inside of fuchsia.

## Related Issues

fxb/55746

## Tests

Tested by rolling the changes manually into the fuchsia tree and running the workstation product. Change works with the current state of not having the nullability experiment enabled as well as later adding it and writing a simple library which uses nullability annotations.

### CODE REVIEW

File: shell/platform/fuchsia/dart_runner/embedder/builtin.dart [hunk 1]
```diff
@@ -25,7 +24,7 @@ class _Logger {
 }
 
 @pragma('vm:entry-point')
-String _rawScript;
+late String _rawScript;
```
Comment (jonahwilliams) [inline shell/platform/fuchsia/dart_runner/embedder/builtin.dart:27]: What is the distinction between the `late` vm entrypoint fields and the nullable ones, is this guaranteed to be set, while the others might still be null?

Comment (chaselatta) [reply shell/platform/fuchsia/dart_runner/embedder/builtin.dart:27]: Yeah... all of the other entry points could be null since they are set by the engine and do have real reasons for being null. The view ref, for example, is only available for things that need a UI. The other entry points will cause the program to not function well when they are null be we currently rely on them being null for host side tests.

I couldn't find a valid reason why this value would be null. If there is one I can update this to be `?` instead of `late`.

Comment (chaselatta) [reply shell/platform/fuchsia/dart_runner/embedder/builtin.dart:27]: I guess it comes down to what we want to do with the `_scriptUri` function below. Is there a valid default if this value is null?

Comment (jonahwilliams) [reply shell/platform/fuchsia/dart_runner/embedder/builtin.dart:27]: Probably no reasonable default, so if it is always going to be set then late is fine.

File: shell/platform/fuchsia/dart-pkg/zircon/lib/src/system.dart [hunk 2]
```diff
@@ -63,23 +62,31 @@ class HandlePairResult extends _Result {
...
+  final List<Handle>? handles;
   @pragma('vm:entry-point')
   const ReadResult(final int status, [this.bytes, this.numBytes, this.handles])
       : super(status);
-  Uint8List bytesAsUint8List() =>
-      bytes.buffer.asUint8List(bytes.offsetInBytes, numBytes);
-  String bytesAsUTF8String() => utf8.decode(bytesAsUint8List());
+
+  Uint8List? bytesAsUint8List() {
+    if (bytes == null) {
```
Comment (alexmarkov) [inline shell/platform/fuchsia/dart-pkg/zircon/lib/src/system.dart:73]: Previously, bytesAsUint8List() would throw an exception if bytes == null.
Maybe we should return non-nullable Uint8List and remove this check to keep old behavior.

File: shell/platform/fuchsia/dart-pkg/zircon/lib/src/system.dart [hunk 3]
```diff
@@ -63,23 +62,31 @@ class HandlePairResult extends _Result {
...
+      return null;
+    }
+    return bytes!.buffer.asUint8List(bytes!.offsetInBytes, numBytes!);
+  }
+
+
+  String? bytesAsUTF8String() => bytes == null ? null : utf8.decode(bytesAsUint8List()!);
```
Comment (alexmarkov) [inline shell/platform/fuchsia/dart-pkg/zircon/lib/src/system.dart:80]: Ditto - consider returning non-nullable String and removing the test for null.

File: shell/platform/fuchsia/dart-pkg/zircon/lib/src/system.dart [hunk 4]
```diff
@@ -88,7 +95,7 @@ class WriteResult extends _Result {
 
 @pragma('vm:entry-point')
 class GetSizeResult extends _Result {
-  final int size;
+  final int? size;
```
Comment (alexmarkov) [inline shell/platform/fuchsia/dart-pkg/zircon/lib/src/system.dart:98]: This migration to null safety looks good.

However, if it is idiomatic to check status before using other fields of the result classes, then we might also consider introducing getter which would return non-nullable value, as it is much easier to work with non-nullable values. Something along the lines:

```dart
class GetSizeResult extends _Result {
  final int? _size;
  int get size => _size!;

  @pragma('vm:entry-point')
  const GetSizeResult(final int status, [this._size]) : super(status);

  @override
  String toString() => 'GetSizeResult(status=$status, size=$_size)';
}
```

This can be a breaking change, so it might make sense to do this in a separate PR.

Comment (chaselatta) [reply shell/platform/fuchsia/dart-pkg/zircon/lib/src/system.dart:98]: That is actually a really good point. Right now folks would have to check status or else they would crash when they tried to check the size. This would probably be the correct time to migrate this.

File: shell/platform/fuchsia/dart-pkg/zircon/lib/src/system.dart (hunk 3 above)
Comment (chaselatta) [reply shell/platform/fuchsia/dart-pkg/zircon/lib/src/system.dart:80]: done.

File: shell/platform/fuchsia/dart-pkg/zircon/lib/src/system.dart (hunk 4 above)
Comment (chaselatta) [reply shell/platform/fuchsia/dart-pkg/zircon/lib/src/system.dart:98]: Updated.
